# Предварительный обзор данных

In [3]:
import pandas
features = pandas.read_csv('./features.csv', index_col='match_id')

In [4]:
features.head()

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
0         1430198770           7       11         5   2098     1489     20   
1         1430220345           0       42         4   1188     1033      9   
2         1430227081           7       33         4   1319     1270     22   
3         1430263531           1       29         4   1779     1056     14   
4         1430282290           7       13         4   1431     1090      8   

          r1_kills  r1_deaths  r1_items  ...  dire_boots_count  \
match_id                                 ...                     
0                0          0         7  ...                 4   
1                0          1        12  ...                 4   
2                0          0        12  ...                 4   
3                0          0         5  ...                 4   
4                1          0         8  ...                 3   

          dire_ward_observer_count  dire_ward_sentry_count  \
match_id                                                     
0                                2                       2   
1                                3                       1   
2                                3                       1   
3                                2                       0   
4                                3                       0   

          dire_first_ward_time  duration  radiant_win  tower_status_radiant  \
match_id                                                                      
0                        -52.0      2874            1                  1796   
1                         -5.0      2463            1                  1974   
2                         13.0      2130            0                     0   
3                         27.0      1459            0                  1920   
4                        -16.0      2449            0                     4   

          tower_status_dire  barracks_status_radiant  barracks_status_dire  
match_id                                                                    
0                         0                       51                     0  
1                         0                       63                     1  
2                      1830                        0                    63  
3                      2047                       50                    63  
4                      1974                        3                    63  

[5 rows x 108 columns]

In [5]:
desc = features.describe()
desc

start_time    lobby_type       r1_hero      r1_level         r1_xp  \
count  9.723000e+04  97230.000000  97230.000000  97230.000000  97230.000000   
mean   1.444232e+09      2.630999     51.517104      3.442672   1233.405801   
std    5.515393e+06      2.835761     32.564211      1.111741    566.588895   
min    1.430199e+09      0.000000      1.000000      0.000000      0.000000   
25%    1.440815e+09      1.000000     22.000000      3.000000    767.000000   
50%    1.446338e+09      1.000000     50.000000      3.000000   1175.000000   
75%    1.448829e+09      7.000000     75.000000      4.000000   1704.000000   
max    1.450313e+09      7.000000    112.000000      6.000000   3319.000000   

            r1_gold         r1_lh      r1_kills     r1_deaths      r1_items  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean    1147.899702     11.231996      0.357009      0.362285      8.271315   
std      464.111662      9.041620      0.663889      0.626704      2.497575   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%      746.000000      2.000000      0.000000      0.000000      7.000000   
50%     1113.000000     11.000000      0.000000      0.000000      8.000000   
75%     1479.000000     19.000000      1.000000      1.000000     10.000000   
max     4332.000000     47.000000      8.000000      5.000000     34.000000   

       ...  dire_boots_count  dire_ward_observer_count  \
count  ...      97230.000000              97230.000000   
mean   ...          3.349553                  2.448339   
std    ...          1.155609                  0.813459   
min    ...          0.000000                  0.000000   
25%    ...          3.000000                  2.000000   
50%    ...          3.000000                  2.000000   
75%    ...          4.000000                  3.000000   
max    ...          9.000000                  9.000000   

       dire_ward_sentry_count  dire_first_ward_time      duration  \
count            97230.000000          95404.000000  97230.000000   
mean                 0.689119             -6.901922   2332.247886   
std                  0.710122             40.701397    715.806850   
min                  0.000000            -84.000000    900.000000   
25%                  0.000000            -31.000000   1818.000000   
50%                  1.000000            -16.000000   2268.000000   
75%                  1.000000              8.000000   2778.000000   
max                 13.000000            300.000000   8452.000000   

        radiant_win  tower_status_radiant  tower_status_dire  \
count  97230.000000          97230.000000       97230.000000   
mean       0.518503           1309.227790        1286.310820   
std        0.499660            853.921365         851.009148   
min        0.000000              0.000000           0.000000   
25%        0.000000             36.000000         256.000000   
50%        1.000000           1824.000000        1798.000000   
75%        1.000000           1974.000000        1974.000000   
max        1.000000           2047.000000        2047.000000   

       barracks_status_radiant  barracks_status_dire  
count             97230.000000          97230.000000  
mean                 40.599095             41.337036  
std                  27.871645             27.064873  
min                   0.000000              0.000000  
25%                   3.000000              3.000000  
50%                  63.000000             60.000000  
75%                  63.000000             63.000000  
max                  63.000000             63.000000  

[8 rows x 108 columns]

#### Описание признаков в таблице

- `match_id`: идентификатор матча в наборе данных
- `start_time`: время начала матча (unixtime)
- `lobby_type`: тип комнаты, в которой собираются игроки (расшифровка в `dictionaries/lobbies.csv`)
- Наборы признаков для каждого игрока (игроки команды Radiant — префикс `rN`, Dire — `dN`):
    - `r1_hero`: герой игрока (расшифровка в dictionaries/heroes.csv)
    - `r1_level`: максимальный достигнутый уровень героя (за первые 5 игровых минут)
    - `r1_xp`: максимальный полученный опыт
    - `r1_gold`: достигнутая ценность героя
    - `r1_lh`: число убитых юнитов
    - `r1_kills`: число убитых игроков
    - `r1_deaths`: число смертей героя
    - `r1_items`: число купленных предметов
- Признаки события "первая кровь" (first blood). Если событие "первая кровь" не успело произойти за первые 5 минут, то признаки принимают пропущенное значение
    - `first_blood_time`: игровое время первой крови
    - `first_blood_team`: команда, совершившая первую кровь (0 — Radiant, 1 — Dire)
    - `first_blood_player1`: игрок, причастный к событию
    - `first_blood_player2`: второй игрок, причастный к событию
- Признаки для каждой команды (префиксы `radiant_` и `dire_`)
    - `radiant_bottle_time`: время первого приобретения командой предмета "bottle"
    - `radiant_courier_time`: время приобретения предмета "courier" 
    - `radiant_flying_courier_time`: время приобретения предмета "flying_courier" 
    - `radiant_tpscroll_count`: число предметов "tpscroll" за первые 5 минут
    - `radiant_boots_count`: число предметов "boots"
    - `radiant_ward_observer_count`: число предметов "ward_observer"
    - `radiant_ward_sentry_count`: число предметов "ward_sentry"
    - `radiant_first_ward_time`: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля
- Итог матча (данные поля отсутствуют в тестовой выборке, поскольку содержат информацию, выходящую за пределы первых 5 минут матча)
    - `duration`: длительность
    - `radiant_win`: 1, если победила команда Radiant, 0 — иначе
    - Состояние башен и барраков к концу матча (см. описание полей набора данных)
        - `tower_status_radiant`
        - `tower_status_dire`
        - `barracks_status_radiant`
        - `barracks_status_dire`

## Признаки с пропусками

In [13]:
df_len = len(features)
desc_count = desc.T['count']
na_counts = desc_count[desc_count < df_len]
na_counts.sort_values()
na_counts.apply(lambda a: (df_len - a) / df_len)

first_blood_time               0.201100
first_blood_team               0.201100
first_blood_player1            0.201100
first_blood_player2            0.452402
radiant_bottle_time            0.161380
radiant_courier_time           0.007117
radiant_flying_courier_time    0.282619
radiant_first_ward_time        0.018883
dire_bottle_time               0.166029
dire_courier_time              0.006953
dire_flying_courier_time       0.268415
dire_first_ward_time           0.018780
Name: count, dtype: float64

##### Какие признаки имеют пропуски среди своих значений? Что могут означать пропуски в этих признаках (ответьте на этот вопрос для двух любых признаков)?

* first_blood_time: игровое время первой крови - первой крови нет в первые 5 минут игры в 20% случаев
* first_blood_team: команда, совершившая первую кровь (0 — Radiant, 1 — Dire) - первой крови нет в первые 5 минут игры в 20% случаев
* first_blood_player1: игрок, причастный к событию - первой крови нет в первые 5 минут игры в 20% случаев
* first_blood_player2: второй игрок, причастный к событию - первое убийство, совершенное одним игроком, происходит в 45% случаев

* radiant_bottle_time: время первого приобретения командой предмета "bottle" - в 16% случаев команда не приобретает предмет "bottle" в первые 5 минут игры
* radiant_courier_time: время приобретения предмета "courier" - в 0.5% случаев команда не приобретает предмет "courier" в первые 5 минут игры
* radiant_flying_courier_time: время приобретения предмета "flying_courier" - в 28% случаев команда не покупает "flying_courier" в первые 5 минут игры
* radiant_first_ward_time: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля - в ~2% случаев команда не ставит "наблюдателя" в первые 5 минут игры

* dire_bottle_time: время первого приобретения командой предмета "bottle" - в 16% случаев команда не приобретает предмет "bottle" в первые 5 минут игры
* dire_courier_time: время приобретения предмета "courier" - в 0.5% случаев команда не приобретает предмет "courier" в первые 5 минут игры
* dire_flying_courier_time: время приобретения предмета "flying_courier" - в 26% случаев команда не покупает "flying_courier" в первые 5 минут игры
* dire_first_ward_time: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля - в ~2% случаев команда не ставит "наблюдателя" в первые 5 минут игры